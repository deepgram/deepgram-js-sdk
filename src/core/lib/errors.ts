export class DeepgramError extends Error {
  protected __dgError = true;

  constructor(message: string) {
    super(message);
    this.name = "DeepgramError";
  }
}

export function isDeepgramError(error: unknown): error is DeepgramError {
  return typeof error === "object" && error !== null && "__dgError" in error;
}

export class DeepgramApiError extends DeepgramError {
  status: number;

  constructor(message: string, status: number) {
    super(message);
    this.name = "DeepgramApiError";
    this.status = status;
  }

  toJSON() {
    return {
      name: this.name,
      message: this.message,
      status: this.status,
    };
  }
}

export class DeepgramUnknownError extends DeepgramError {
  originalError: unknown;

  constructor(message: string, originalError: unknown) {
    super(message);
    this.name = "DeepgramUnknownError";
    this.originalError = originalError;
  }
}

export class DeepgramVersionError extends DeepgramError {
  constructor() {
    super(
      `You are attempting to use an old format for a newer SDK version. Read more here: https://dpgr.am/js-v3`
    );

    this.name = "DeepgramVersionError";
  }
}

/**
 * Enhanced WebSocket error that captures additional debugging information
 * including status codes, request IDs, and response headers when available.
 */
export class DeepgramWebSocketError extends DeepgramError {
  originalEvent?: ErrorEvent | Event;
  statusCode?: number;
  requestId?: string;
  responseHeaders?: Record<string, string>;
  url?: string;
  readyState?: number;

  constructor(
    message: string,
    options: {
      originalEvent?: ErrorEvent | Event;
      statusCode?: number;
      requestId?: string;
      responseHeaders?: Record<string, string>;
      url?: string;
      readyState?: number;
    } = {}
  ) {
    super(message);
    this.name = "DeepgramWebSocketError";
    this.originalEvent = options.originalEvent;
    this.statusCode = options.statusCode;
    this.requestId = options.requestId;
    this.responseHeaders = options.responseHeaders;
    this.url = options.url;
    this.readyState = options.readyState;
  }

  toJSON() {
    return {
      name: this.name,
      message: this.message,
      statusCode: this.statusCode,
      requestId: this.requestId,
      responseHeaders: this.responseHeaders,
      url: this.url,
      readyState: this.readyState,
      originalEvent: this.originalEvent
        ? {
            type: this.originalEvent.type,
            timeStamp: this.originalEvent.timeStamp,
          }
        : undefined,
    };
  }
}
