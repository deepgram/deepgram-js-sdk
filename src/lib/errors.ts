export class DeepgramError extends Error {
  protected __dgError = true;

  constructor(message: string) {
    super(message);
    this.name = "DeepgramError";
  }
}

export function isDeepgramError(error: unknown): error is DeepgramError {
  return typeof error === "object" && error !== null && "__dgError" in error;
}

export class DeepgramApiError extends DeepgramError {
  status: number;

  constructor(message: string, status: number) {
    super(message);
    this.name = "DeepgramApiError";
    this.status = status;
  }

  toJSON() {
    return {
      name: this.name,
      message: this.message,
      status: this.status,
    };
  }
}

export class DeepgramUnknownError extends DeepgramError {
  originalError: unknown;

  constructor(message: string, originalError: unknown) {
    super(message);
    this.name = "DeepgramUnknownError";
    this.originalError = originalError;
  }
}

export class DeepgramVersionError extends DeepgramError {
  constructor() {
    super(
      `You are attempting to use an old format for a newer SDK version. Read more here: https://dpgr.am/js-v3`
    );

    this.name = "DeepgramVersionError";
  }
}
