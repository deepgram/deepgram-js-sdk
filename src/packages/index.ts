export { AbstractClient } from "./AbstractClient";
export { AbstractRestfulClient } from "./AbstractRestfulClient";
export { AbstractWsClient } from "./AbstractWsClient";
export { ListenClient } from "./ListenClient";
export { LiveClient } from "./LiveClient";
export { ManageClient } from "./ManageClient";
export { OnPremClient } from "./OnPremClient";
export { PrerecordedClient } from "./PrerecordedClient";
