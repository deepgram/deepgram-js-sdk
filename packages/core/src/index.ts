import { DeepgramClientOptions, IKeyFactory } from "./lib/types/DeepgramClientOptions";
import { DeepgramVersionError } from "./lib/errors";
import DeepgramClient from "./DeepgramClient";

/**
 * This class is deprecated and should not be used. It throws a `DeepgramVersionError` when instantiated.
 *
 * @deprecated
 * @see https://dpgr.am/js-v3
 */
class Deepgram {
  constructor(protected apiKey: string, protected apiUrl?: string, protected requireSSL?: boolean) {
    throw new DeepgramVersionError();
  }
}

/**
 * Creates a new Deepgram client instance.
 *
 * @param {DeepgramClientArgs} args - Arguments to pass to the Deepgram client constructor.
 * @returns A new Deepgram client instance.
 */
function createClient(): DeepgramClient;
function createClient(key?: string | IKeyFactory): DeepgramClient;
function createClient(options?: DeepgramClientOptions): DeepgramClient;
function createClient(key?: string | IKeyFactory, options?: DeepgramClientOptions): DeepgramClient;
function createClient(
  keyOrOptions?: string | IKeyFactory | DeepgramClientOptions,
  options?: DeepgramClientOptions
): DeepgramClient {
  let resolvedOptions: DeepgramClientOptions = {};

  if (typeof keyOrOptions === "string" || typeof keyOrOptions === "function") {
    if (typeof options === "object") {
      resolvedOptions = options;
    }

    resolvedOptions.key = keyOrOptions;
  } else if (typeof keyOrOptions === "object") {
    resolvedOptions = keyOrOptions;
  }

  return new DeepgramClient(resolvedOptions);
}

export { createClient, DeepgramClient, Deepgram };

/**
 * Helpful exports.
 */
export * from "./packages";
export * from "./lib/types";
export * from "./lib/enums";
export * from "./lib/constants";
export * from "./lib/errors";
export * from "./lib/helpers";

/**
 * Captions. These will be tree-shaken if unused.
 *
 * @see https://github.com/deepgram/deepgram-node-captions
 *
 * import/export declarations don't do anything but set up an alias to the
 * exported variable, they do not count as a "use". Given their semantics,
 * they are tracked specially by any bundler and will not adversely affect
 * tree-shaking.
 */
export { webvtt, srt } from "@deepgram/captions";
