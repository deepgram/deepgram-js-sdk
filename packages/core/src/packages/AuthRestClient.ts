import { isDeepgramError } from "../lib/errors";
import type { DeepgramResponse } from "../lib/types/DeepgramResponse";
import type { GrantTokenResponse } from "../lib/types/GrantTokenResponse";
import type { GrantTokenSchema } from "../lib/types/GrantTokenSchema";
import { AbstractRestClient } from "./AbstractRestClient";

export class AuthRestClient extends AbstractRestClient {
  public namespace: string = "auth";

  /**
   * Generates a new temporary token for the Deepgram API.
   * @param options Optional configuration options for the token generation. Includes ttl_seconds to set token expiration.
   * @param endpoint Optional custom endpoint to use for the request. Defaults to ":version/auth/grant".
   * @returns Object containing the result of the request or an error if one occurred. Result will contain access_token and expires_in properties.
   */
  public async grantToken(
    options: GrantTokenSchema = {},
    endpoint = ":version/auth/grant"
  ): Promise<DeepgramResponse<GrantTokenResponse>> {
    try {
      const requestUrl = this.getRequestUrl(endpoint);
      const body = JSON.stringify(options);
      const result: GrantTokenResponse = await this.post(requestUrl, body, {
        headers: { "Content-Type": "application/json" },
      }).then((result) => result.json());

      return { result, error: null };
    } catch (error) {
      if (isDeepgramError(error)) {
        return { result: null, error };
      }

      throw error;
    }
  }
}
