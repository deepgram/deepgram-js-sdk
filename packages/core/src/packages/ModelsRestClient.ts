import { isDeepgramError } from "../lib/errors";
import {
  DeepgramResponse,
  GetModelResponse,
  GetModelsResponse,
  GetModelsSchema,
} from "../lib/types";
import { AbstractRestClient } from "./AbstractRestClient";

/**
 * Represents a REST client for interacting with the Deepgram API.
 *
 * The `ModelsRestClient` class provides methods for interacting with the Deepgram API to retrieve information about available models.
 * @extends AbstractRestClient
 */
export class ModelsRestClient extends AbstractRestClient {
  public namespace: string = "models";

  /**
   * Retrieves a list of all available models.
   *
   * @param endpoint - (optional) The endpoint to request.
   * @returns A promise that resolves with the response from the Deepgram API.
   * @example
   * ```typescript
   * import { createClient } from "@deepgram/sdk";
   *
   * const deepgram = createClient(DEEPGRAM_API_KEY);
   * const { result: models, error } = deepgram.models.getAll();
   *
   * if (error) {
   *   console.error(error);
   * } else {
   *   console.log(models);
   * }
   * ```
   */
  async getAll(
    endpoint = ":version/models",
    options: GetModelsSchema = {}
  ): Promise<DeepgramResponse<GetModelsResponse>> {
    try {
      const requestUrl = this.getRequestUrl(endpoint, {}, options);
      const result: GetModelsResponse = await this.get(requestUrl).then((result) => result.json());

      return { result, error: null };
    } catch (error) {
      if (isDeepgramError(error)) {
        return { result: null, error };
      }

      throw error;
    }
  }

  /**
   * Retrieves information about a specific model.
   *
   * @param modelId - The UUID of the model to retrieve.
   * @param endpoint - (optional) The endpoint to request.
   * @returns A promise that resolves with the response from the Deepgram API.
   * @example
   * ```typescript
   * import { createClient } from "@deepgram/sdk";
   *
   * const deepgram = createClient(DEEPGRAM_API_KEY);
   * const { result: model, error } = deepgram.models.getModel("modelId");
   *
   * if (error) {
   *   console.error(error);
   * } else {
   *   console.log(model);
   * }
   * ```
   */
  async getModel(
    modelId: string,
    endpoint = ":version/models/:modelId"
  ): Promise<DeepgramResponse<GetModelResponse>> {
    try {
      const requestUrl = this.getRequestUrl(endpoint, { modelId });
      const result: GetModelResponse = await this.get(requestUrl).then((result) => result.json());

      return { result, error: null };
    } catch (error) {
      if (isDeepgramError(error)) {
        return { result: null, error };
      }

      throw error;
    }
  }
}
