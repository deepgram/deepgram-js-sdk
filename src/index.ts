import DeepgramClient from "./DeepgramClient";
import { DeepgramVersionError } from "./lib/errors";
import type { DeepgramClientOptions } from "./lib/types";

/**
 * Major version fallback error
 */
class Deepgram {
  constructor(protected apiKey: string, protected apiUrl?: string, protected requireSSL?: boolean) {
    throw new DeepgramVersionError();
  }
}

/**
 * Creates a new Deepgram Client.
 */
const createClient = (apiKey: string, options: DeepgramClientOptions = {}): DeepgramClient => {
  return new DeepgramClient(apiKey, options);
};

export { createClient, DeepgramClient, Deepgram };

/**
 * Helpful exports.
 */
export * from "./packages";
export * from "./lib/types";
export * from "./lib/enums";
export * from "./lib/constants";
export * from "./lib/errors";
export * from "./lib/helpers";

/**
 * Captions. These will be tree-shaken if unused.
 *
 * @see https://github.com/deepgram/deepgram-node-captions
 *
 * import/export declarations don't do anything but set up an alias to the
 * exported variable, they do not count as a "use". Given their semantics,
 * they are tracked specially by any bundler and will not adversely affect
 * tree-shaking.
 */
export { webvtt, srt } from "@deepgram/captions";
