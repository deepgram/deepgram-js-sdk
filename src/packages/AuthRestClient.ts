import { isDeepgramError } from "../lib/errors";
import type { DeepgramResponse } from "../lib/types/DeepgramResponse";
import type { GrantTokenResponse } from "../lib/types/GrantTokenResponse";
import { AbstractRestClient } from "./AbstractRestClient";

export class AuthRestClient extends AbstractRestClient {
  public namespace: string = "auth";

  /**
   * Generates a new temporary token for the Deepgram API.
   * @param endpoint Optional custom endpoint to use for the request. Defaults to ":version/auth/grant".
   * @returns Object containing the result of the request or an error if one occurred. Result will contain access_token and expires_in properties.
   */
  public async grantToken(
    endpoint = ":version/auth/grant"
  ): Promise<DeepgramResponse<GrantTokenResponse>> {
    try {
      const requestUrl = this.getRequestUrl(endpoint);
      const result: GrantTokenResponse = await this.post(requestUrl, "").then((result) =>
        result.json()
      );

      return { result, error: null };
    } catch (error) {
      if (isDeepgramError(error)) {
        return { result: null, error };
      }

      throw error;
    }
  }
}
