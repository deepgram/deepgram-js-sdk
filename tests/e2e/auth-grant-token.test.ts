import { createClient } from "../../src/index";
import { AuthRestClient } from "../../src/packages/AuthRestClient";
import { DeepgramError } from "../../src/lib/errors";
import { structureOnlySerializer, setupApiMocks, cleanupApiMocks } from "../__utils__";

describe("auth grantToken E2E", () => {
  let deepgram: ReturnType<typeof createClient>;

  beforeAll(() => {
    // Set up API mocks (only active when not updating snapshots)
    setupApiMocks();

    // Use mock API key when mocking, real one when updating snapshots
    const apiKey = process.env.DEEPGRAM_API_KEY || "mock-api-key";
    deepgram = createClient(apiKey);

    // Add our custom serializer
    expect.addSnapshotSerializer(structureOnlySerializer);
  });

  afterAll(() => {
    // Clean up mocks
    cleanupApiMocks();
  });

  it("should generate a temporary access token", async () => {
    const { result, error } = await deepgram.auth.grantToken();

    // Verify no error occurred
    expect(error).toBeNull();
    expect(result).toBeTruthy();

    // Type guard to ensure result is not null for subsequent operations
    if (!result) {
      throw new Error("Result should not be null after toBeTruthy check");
    }

    // Test the structure with snapshot
    expect(result).toMatchSnapshot("auth-grantToken-response-structure");

    // Essential structural validation
    expect(result).toHaveProperty("access_token");
    expect(result).toHaveProperty("expires_in");

    // Verify data types and format
    expect(typeof result.access_token).toBe("string");
    expect(typeof result.expires_in).toBe("number");

    // Verify access token is a non-empty string (should be JWT format in real API)
    expect(result.access_token.length).toBeGreaterThan(0);

    // Verify expires_in is a positive number (typically in seconds)
    expect(result.expires_in).toBeGreaterThan(0);
  }, 30000);

  it("should handle custom endpoint for token generation", async () => {
    const customEndpoint = ":version/auth/grant";
    const { result, error } = await deepgram.auth.grantToken(customEndpoint);

    // Verify no error occurred
    expect(error).toBeNull();
    expect(result).toBeTruthy();

    if (!result) {
      throw new Error("Result should not be null after toBeTruthy check");
    }

    // Verify the response structure is consistent
    expect(result).toHaveProperty("access_token");
    expect(result).toHaveProperty("expires_in");
    expect(typeof result.access_token).toBe("string");
    expect(typeof result.expires_in).toBe("number");
  }, 30000);

  it("should handle DeepgramError and return it in response", async () => {
    const apiKey = process.env.DEEPGRAM_API_KEY || "mock-api-key";
    const authClient = new AuthRestClient({ key: apiKey });

    // Mock the post method to throw a DeepgramError
    const mockError = new DeepgramError("Test DeepgramError");
    jest.spyOn(authClient as any, "post").mockRejectedValue(mockError);

    const { result, error } = await authClient.grantToken();

    expect(result).toBeNull();
    expect(error).toBe(mockError);
    expect(error).toBeInstanceOf(DeepgramError);

    // Clean up spy
    jest.restoreAllMocks();
  });

  it("should re-throw non-DeepgramError", async () => {
    const apiKey = process.env.DEEPGRAM_API_KEY || "mock-api-key";
    const authClient = new AuthRestClient({ key: apiKey });

    // Mock the post method to throw a regular Error
    const mockError = new Error("Network error");
    jest.spyOn(authClient as any, "post").mockRejectedValue(mockError);

    await expect(authClient.grantToken()).rejects.toThrow("Network error");
    await expect(authClient.grantToken()).rejects.toThrow(Error);

    // Clean up spy
    jest.restoreAllMocks();
  });
});
