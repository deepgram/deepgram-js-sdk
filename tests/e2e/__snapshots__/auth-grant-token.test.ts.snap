// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth grantToken E2E should generate a temporary access token: auth-grantToken-response-structure 1`] = `
{
  "access_token": "<string>",
  "expires_in": "<number>"
}
`;
